File: apps/dawnchat-android/android/app/src/main/java/com/dawnchat/app/plugins/DawnTtsPlugin.kt
package com.dawnchat.app.plugins

import android.util.Log
import com.dawnchat.app.DawnChatApp
import com.dawnchat.app.R
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.github.jan.supabase.auth.auth
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * POST dawn-tts Edge function with Supabase session token; writes audio/mpeg to cache and returns absolute path.
 * Token never crosses into JS.
 */
@CapacitorPlugin(name = "DawnTts")
class DawnTtsPlugin : Plugin() {

    private val httpClient: OkHttpClient =
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(90, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build()

    @PluginMethod
    fun synthesizeToFile(call: PluginCall) {
        val text = call.getString("text")?.trim().orEmpty()
        if (text.isEmpty()) {
            call.reject("Field text is required", "text_required")
            return
        }

        val voice = call.getString("voice")?.trim()
        val rate = call.getString("rate")?.trim()
        val volume = call.getString("volume")?.trim()
        val pitch = call.getString("pitch")?.trim()

        val app = activity.applicationContext as? DawnChatApp
        val client = app?.supabaseClientOrNull()
        val session = client?.auth?.currentSessionOrNull()
        val accessToken = session?.accessToken?.trim().orEmpty()
        if (accessToken.isEmpty()) {
            call.reject("Not signed in or session has no access token", "not_authenticated")
            return
        }

        val baseUrl = activity.getString(R.string.dawnchat_supabase_url).trim().trimEnd('/')
        val anonKey = activity.getString(R.string.dawnchat_supabase_anon_key).trim()
        if (baseUrl.isEmpty() || anonKey.isEmpty()) {
            call.reject("Supabase URL or anon key is not configured", "supabase_config_missing")
            return
        }

        val url = "$baseUrl/functions/v1/dawn-tts"
        val bodyJson = JSONObject()
        bodyJson.put("text", text)
        if (!voice.isNullOrEmpty()) bodyJson.put("voice", voice)
        if (!rate.isNullOrEmpty()) bodyJson.put("rate", rate)
        if (!volume.isNullOrEmpty()) bodyJson.put("volume", volume)
        if (!pitch.isNullOrEmpty()) bodyJson.put("pitch", pitch)

        val mediaType = JSON_MEDIA
        val requestBody = bodyJson.toString().toRequestBody(mediaType)
        val request =
            Request.Builder()
                .url(url)
                .post(requestBody)
                .header("Authorization", "Bearer $accessToken")
                .header("apikey", anonKey)
                .header("Content-Type", "application/json")
                .header("Accept", "audio/mpeg")
                .build()

        Thread {
            try {
                httpClient.newCall(request).execute().use { response ->
                    val bytes = response.body.bytes()
                    if (!response.isSuccessful) {
                        val parsed = parseErrorJson(bytes)
                        val code = parsed.first ?: "dawn_tts_http_${response.code}"
                        val message = parsed.second ?: response.message.ifEmpty { "HTTP ${response.code}" }
                        activity.runOnUiThread { call.reject(message, code) }
                        return@use
                    }

                    val contentType = response.header("Content-Type")?.lowercase().orEmpty()
                    if (contentType.contains("application/json")) {
                        activity.runOnUiThread {
                            call.reject("Unexpected JSON response; ensure Accept is audio/mpeg", "unexpected_json")
                        }
                        return@use
                    }

                    val dir = File(activity.cacheDir, "dawn-tts").apply { mkdirs() }
                    val outFile = File(dir, "${UUID.randomUUID()}.mp3")
                    outFile.writeBytes(bytes)

                    val ret = JSObject()
                    ret.put("path", outFile.absolutePath)
                    activity.runOnUiThread { call.resolve(ret) }
                }
            } catch (e: Exception) {
                Log.e(TAG, "synthesizeToFile failed", e)
                activity.runOnUiThread {
                    call.reject(e.message ?: "request failed", "dawn_tts_request_failed", e)
                }
            }
        }.start()
    }

    private fun parseErrorJson(bytes: ByteArray): Pair<String?, String?> {
        if (bytes.isEmpty()) return null to null
        return try {
            val jo = JSONObject(bytes.decodeToString())
            val code = jo.optString("code", "").trim().takeIf { it.isNotEmpty() }
            val message = jo.optString("message", "").trim().takeIf { it.isNotEmpty() }
            code to message
        } catch (_: Exception) {
            null to null
        }
    }

    companion object {
        private const val TAG = "DawnTtsPlugin"
        private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()
    }
}

File: apps/dawnchat-ios/ios/App/App/DawnTtsPlugin.swift
import Capacitor
import Foundation

/// POST dawn-tts with Supabase session; writes MP3 to caches and returns path. Token stays native.
@objc(DawnTtsPlugin)
public class DawnTtsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DawnTtsPlugin"
    public let jsName = "DawnTts"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "synthesizeToFile", returnType: CAPPluginReturnPromise),
    ]

    @objc func synthesizeToFile(_ call: CAPPluginCall) {
        guard let text = call.getString("text")?.trimmingCharacters(in: .whitespacesAndNewlines), !text.isEmpty else {
            call.reject("Field text is required", "text_required")
            return
        }

        Task {
            do {
                let token = try await MobileAuthRepository.shared.dawnTtsAccessToken()
                let bundleUrl = (Bundle.main.object(forInfoDictionaryKey: "DawnChatSupabaseURL") as? String)?
                    .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                let anon = (Bundle.main.object(forInfoDictionaryKey: "DawnChatSupabaseAnonKey") as? String)?
                    .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
                let base = bundleUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
                if base.isEmpty || anon.isEmpty {
                    await MainActor.run {
                        call.reject("Supabase URL or anon key is not configured", "supabase_config_missing")
                    }
                    return
                }
                guard let url = URL(string: base + "/functions/v1/dawn-tts") else {
                    await MainActor.run { call.reject("Invalid Supabase URL", "invalid_url") }
                    return
                }

                var body: [String: Any] = ["text": text]
                if let v = call.getString("voice")?.trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty {
                    body["voice"] = v
                }
                if let r = call.getString("rate")?.trimmingCharacters(in: .whitespacesAndNewlines), !r.isEmpty {
                    body["rate"] = r
                }
                if let v = call.getString("volume")?.trimmingCharacters(in: .whitespacesAndNewlines), !v.isEmpty {
                    body["volume"] = v
                }
                if let p = call.getString("pitch")?.trimmingCharacters(in: .whitespacesAndNewlines), !p.isEmpty {
                    body["pitch"] = p
                }
                let jsonData = try JSONSerialization.data(withJSONObject: body)

                var req = URLRequest(url: url)
                req.httpMethod = "POST"
                req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                req.setValue(anon, forHTTPHeaderField: "apikey")
                req.setValue("application/json", forHTTPHeaderField: "Content-Type")
                req.setValue("audio/mpeg", forHTTPHeaderField: "Accept")
                req.httpBody = jsonData
                req.timeoutInterval = 90

                let (data, response) = try await URLSession.shared.data(for: req)
                guard let http = response as? HTTPURLResponse else {
                    await MainActor.run { call.reject("Invalid response", "dawn_tts_request_failed") }
                    return
                }
                if !(200 ... 299).contains(http.statusCode) {
                    let (code, message) = Self.parseErrorPayload(data: data, status: http.statusCode)
                    await MainActor.run { call.reject(message, code) }
                    return
                }
                let ct = http.value(forHTTPHeaderField: "Content-Type")?.lowercased() ?? ""
                if ct.contains("application/json") {
                    await MainActor.run {
                        call.reject("Unexpected JSON response; ensure Accept is audio/mpeg", "unexpected_json")
                    }
                    return
                }

                let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
                guard let cacheRoot = caches.first else {
                    await MainActor.run { call.reject("No cache directory", "cache_unavailable") }
                    return
                }
                let dir = cacheRoot.appendingPathComponent("dawn-tts", isDirectory: true)
                try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
                let out = dir.appendingPathComponent(UUID().uuidString + ".mp3")
                try data.write(to: out)
                await MainActor.run { call.resolve(["path": out.path]) }
            } catch {
                let ns = error as NSError
                if ns.domain == "DawnTts", let dawnCode = ns.userInfo["dawnCode"] as? String {
                    await MainActor.run { call.reject(ns.localizedDescription, dawnCode) }
                } else {
                    await MainActor.run {
                        call.reject(error.localizedDescription, "dawn_tts_request_failed", error)
                    }
                }
            }
        }
    }

    private static func parseErrorPayload(data: Data, status: Int) -> (String, String) {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return ("dawn_tts_http_\(status)", HTTPURLResponse.localizedString(forStatusCode: status))
        }
        let code = (obj["code"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let message = (obj["message"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
        let c = (code?.isEmpty == false) ? code! : "dawn_tts_http_\(status)"
        let m = (message?.isEmpty == false) ? message! : HTTPURLResponse.localizedString(forStatusCode: status)
        return (c, m)
    }
}
